native/core/src/execution/shuffle/shuffle_writer.rs

```diff
@@ -1567,17 +1585,41 @@ pub fn write_ipc_compressed<W: Write + Seek>(
     let mut timer = ipc_time.timer();
     let start_pos = output.stream_position()?;
 
-    // write ipc_length placeholder
-    output.write_all(&[0u8; 8])?;
+    // seek past ipc_length placeholder
+    output.seek_relative(8)?;
+
+    // write number of columns because JVM side needs to know how many addresses to allocate
+    let field_count = batch.schema().fields().len();
+    output.write_all(&field_count.to_le_bytes())?;
 
     let output = match codec {
         CompressionCodec::None => {
+            output.write_all(b"NONE")?;
             let mut arrow_writer = StreamWriter::try_new(output, &batch.schema())?;
             arrow_writer.write(batch)?;
             arrow_writer.finish()?;
             arrow_writer.into_inner()?
         }
+        CompressionCodec::Snappy => {
+            output.write_all(b"SNAP")?;
+            let mut wtr = snap::write::FrameEncoder::new(output);
+            let mut arrow_writer = StreamWriter::try_new(&mut wtr, &batch.schema())?;
+            arrow_writer.write(batch)?;
+            arrow_writer.finish()?;
+            wtr.into_inner()
+                .map_err(|e| DataFusionError::Execution(format!("lz4 compression error: {}", e)))?
+        }
+        CompressionCodec::Lz4Frame => {
+            output.write_all(b"LZ4_")?;
+            let mut wtr = lz4_flex::frame::FrameEncoder::new(output);
+            let mut arrow_writer = StreamWriter::try_new(&mut wtr, &batch.schema())?;
```

Comment:
probably dumb question, looks like the writers gets created for every input.

In real life example it will be recreated for every `batch`, probably the writer can be reused until the data stream has completed?

Reply:
Spark's approach is to write shuffled output data in blocks where each block is independent (contains schema + data for that batch), so it doesn't allow for a streaming approach. We accumulate data in the shuffle writer until we reach the configured batch size and then write that batch out to bytes. There is also no guarantee of the order in which the shuffle reader will read these blocks, so that is another reason why we cannot keep a writer open for multiple batches.

Reply:
do you think it makes any sense if I play with reusable writers to check if it gives any benefit on multiple bacthes? I feel we can have a map of it where key is schema hash + compression algorithm. During real job runtime its probably not expected to have more than 50-100 writers

Reply:
Feel free to experiment, and let's see what else we can learn. I'm not sure exactly what you are planning on trying but it seems that we could just have one writer per `PartitionBuffer` and write multiple batches to it. It may complicate the spill logic but you can just disable that for testing the idea. I wonder how this  would differ from just increasing the shuffle batch size so that we write larger batches?

Reply:
yeah, thats another way to increase `batch_size` so the compression writer will be created less often.
The test idea was to find out how expensive is the writer creation, to see if this can be an issue.

Perhaps I can just try to create a bench and quantify it so it gives us some ideas.

Reply:
what I found is 
```
Time elapsed in writer_create is: 107.125µs
Time elapsed in total is: 1.699042ms
```
The average for creation is 7% of overall time. However to make writer a singleton is not a trivial task...

Reply:
In the PR that follows this one (https://github.com/apache/datafusion-comet/pull/1190) I propose replacing `ArrowWriter` with a faster proprietary version.